### ISSUE TITLE
Trailing commas for block arguments should be parsed as mlhs

### ISSUE THREAD

**Issue by backus** [whitequark/parser#312]
Ruby lets you use a trailing comma in a block argument for destructuring

```ruby
people = { 'John' => 22, 'Jim' => 23 }

people.each do |name,|
  puts "Hello, #{name}!"
end

# >> Hello, John!
# >> Hello, Jim!
```

Removing the comma results in different behavior:

```ruby
people = { 'John' => 22, 'Jim' => 23 }

people.each do |name|
  puts "Hello, #{name}!"
end

# >> Hello, ["John", 22]!
# >> Hello, ["Jim", 23]!
```

Parser produces the same AST for both cases:

```sh
$ ruby-parse --version
ruby-parse based on parser version 2.3.1.2

$ ruby-parse -e 'foo.each { |bar,| bar }'
(block
  (send
    (send nil :foo) :each)
  (args
    (arg :bar))
  (lvar :bar))

$ ruby-parse -e 'foo.each { |bar| bar }'
(block
  (send
    (send nil :foo) :each)
  (args
    (arg :bar))
  (lvar :bar))
```

## Expected behavior

Parser correctly handles destructuring like this for assignments:

```sh
$ ruby-parse -e 'foo = bar'
(lvasgn :foo
  (send nil :bar))
$ ruby-parse -e 'foo, = bar'
(masgn
  (mlhs
    (lvasgn :foo))
  (send nil :bar))
```

and parser also supports other forms of destructuring for block arguments:

```sh
$ ruby-parse -e 'foo.each { |(bar)| bar }'
(block
  (send
    (send nil :foo) :each)
  (args
    (mlhs
      (arg :bar)))
  (lvar :bar))
```

so it seems like the appropriate fix would be a parse that wraps the block arguments in `(mlhs ...)` like so:

```sh
$ ruby-parse -e 'foo.each { |bar,| bar }'
(block
  (send
    (send nil :foo) :each)
  (args
    (mlhs
      (arg :bar)))
  (lvar :bar))
```

Thanks for your great work on parser! I would be happy to look into this bug later today when I have some extra time.

**Comment by backus** [comment]
Here is a failing test:

```ruby
def test_masgn_block
  assert_parses(
    s(:block,
      s(:send,
        s(:send, s(:self), :foo), :each),
      s(:args,
        s(:mlhs,
          s(:arg, :bar))),
      s(:lvar, :bar)),
    %q(self.foo.each { |(bar)| bar }),
    %q{},
    ALL_VERSIONS - %w(1.8))

  assert_parses(
    s(:block,
      s(:send,
        s(:send, s(:self), :foo), :each),
      s(:args,
        s(:mlhs,
          s(:arg, :bar))),
      s(:lvar, :bar)),
    %q(self.foo.each { |bar,| bar }),
    %q{},
    ALL_VERSIONS - %w(1.8))
end
```

**Comment by mbj** [comment]
@backus did you check if unparser can correctly produce the proposed AST here? Can you open an issue if not?

**Comment by whitequark** [comment]
Thanks for the report. This is indeed a bug. I believe it is in this code: https://github.com/whitequark/parser/blob/master/lib/parser/builders/default.rb#L1031-L1035

**Comment by backus** [comment]
@mbj I think unparser is fine. Given the AST I mentioned it produces the following:

```ruby
=> s(:block,
  s(:send,
    s(:send,
      s(:self), :foo), :each),
  s(:args,
    s(:mlhs,
      s(:arg, :bar))),
  s(:lvar, :bar))
irb(main):024:0> puts Unparser.unparse(block_mlhs)
self.foo.each do |(bar)|
  bar
end
```

which is semantically equivalent.

**Comment by backus** [comment]
@whitequark cool I'll try to take a stab at a fix later

**Comment by whitequark** [comment]
Reopening because of a bug described in https://github.com/whitequark/parser/pull/313#issuecomment-249558060.

**Comment by iliabylich** [comment]
@whitequark It seems that there's some [flag](https://github.com/ruby/ruby/blob/trunk/parse.y#L3374) (`1`) in MRI that indicates that source code has a trailing comma. Do you know how does it work? As I understand, arguments are
```
(req, opt, rest, post, kwargs_and_block)
```
Is it some flag for invisible nameless splat that is ignored by `Method/Proc#arity`?

I hope it's not.

By the way, Ripper has `0` flag for some reason.

``` ruby
Ripper.sexp("p{|a,|}")[1][0][2][1][1]
=> [:params, [[:@ident, "a", [1, 3]]], nil, 0, nil, nil, nil, nil] 
Ripper.sexp("p{|a|}")[1][0][2][1][1]
=> [:params, [[:@ident, "a", [1, 3]]], nil, nil, nil, nil, nil, nil]
```

Also proc with this argument expands its arguments:
``` ruby
proc { |a,| a }.call([1])
 => 1
# Just like:
proc { |a, *| a }.call([1])
 => 1
```

For me it looks like a shortcut for skipping trailing `*`

**Comment by whitequark** [comment]
That seems correct to me. So what we should do in this case is insert a synthesized formal rest argument (with the same location as for synthesized begin...)

**Comment by iliabylich** [comment]
What do you mean by `synthesized formal rest argument `? Introducing new node like `kwbegin`?

**Comment by whitequark** [comment]
No. `kwbegin` is completely unrelated. Actually, this is a real problem because it's unclear what the location should point to... Anyway, for now just make parsers put a `(restarg)` there, I'll think about it more.

**Comment by iliabylich** [comment]
But handling it as a blank restarg breaks proc arity. And the code generated by unparser will not be the same

**Comment by whitequark** [comment]
That's a good point. This is probably the first moment since I started writing parser that I legitimately have no idea what's going on... I'll dig into the MRI source the next weekend and figure out the final answer to this stupid issue.

I'm not sure how much sense it makes for you to do anything before that since fixing it *somehow* is trivial but fixing it *right* appears to be beyond the knowledge of either of us at the moment.

**Comment by mbj** [comment]
@whitequark Your work is generally under-appreciated. Thx for this. I can only see a small glimpse on how painful it must be first hand.

**Comment by whitequark** [comment]
*long sigh*

Ok, I've figured out the mysterious flag. The position in which `0` appears here:

```ruby
Ripper.sexp("p{|a,|}")[1][0][2][1][1]
=> [:params, [[:@ident, "a", [1, 3]]], nil, 0, nil, nil, nil, nil]
```

corresponds to the `rb_args_info->restarg` field. parse.y [stores the raw integer value 1](https://github.com/ruby/ruby/blob/v2_3_1/parse.y#L3411) into it only in the case of "excessed comma" (as the parse.y source calls this pattern). This raw integer value 1 corresponds to a valid fixnum 0 when decoded with the MRI tagged pointer rules.

Then, we can [look at compile.c](https://github.com/ruby/ruby/blob/1b94420174641d78b99c45074e8770a91efa2d2b/compile.c#L1378), where it just sets `last_comma = 1` as a special case, which then [sets `ambiguous_param0` flag](https://github.com/ruby/ruby/blob/1b94420174641d78b99c45074e8770a91efa2d2b/compile.c#L1466-L1469) in the compiled function, which then changes [a bunch of inane fuckery](https://github.com/ruby/ruby/search?utf8=%E2%9C%93&q=ambiguous_param0) all over the compiler that I have zero inclination to reverse-engineer because I'm just doing a parser here.

The bottom line is that the case of "excessed comma" needs to gain its own AST node. Since this is a breaking change, it needs to gain a flag in the AST builder, like [emit_lambda](https://github.com/whitequark/parser/blob/master/lib/parser/builders/default.rb#L8-L20) does. When ran without this flag, I suggest restoring the behavior pre-4d4a2db8. The motivation is as follows: we cannot do the right thing anyway (or we didn't need the flag) and no one complained for years before the fix so it couldn't have been that bad.

@backus @iliabylich Can one of you please prepare a PR?

**Comment by iliabylich** [comment]
@whitequark Any ideas how to name this new node? `s(:excessed_comma)`?

**Comment by backus** [comment]
Wow this is insane. Thanks for taking a deeper look into this @whitequark.

**Comment by alexdowad** [comment]
@whitequark Don't envy you one little bit. Ruby internals are a pain, and the dev team does not seem to have any inclination to stop adding new features and clean up what is already there.

**Comment by whitequark** [comment]
*long sign again* So I was hesitant to merge #322 yesterday, and it turns out, for a good reason. Let's take a closer look at how proc and masgn work:

```
[1] pry(main)> proc{|a,| [a]}.call(1)
=> [1]
[2] pry(main)> proc{|a,| [a]}.call(1, 2)
=> [1]
[3] pry(main)> proc{|a,| [a]}.call(1, 2, 3)
=> [1]
[4] pry(main)> proc{|a,b| [a,b]}.call(1)
=> [1, nil]
[5] pry(main)> proc{|a,b| [a,b]}.call(1, 2)
=> [1, 2]
[6] pry(main)> proc{|a,b| [a,b]}.call(1, 2, 3)
=> [1, 2]
[7] pry(main)> proc{|a,b,| [a,b]}.call(1)
=> [1, nil]
[8] pry(main)> proc{|a,b,| [a,b]}.call(1, 2)
=> [1, 2]
[9] pry(main)> proc{|a,b,| [a,b]}.call(1, 2, 3)
=> [1, 2]
[10] pry(main)> 
[11] pry(main)> a, = 1; p [a]
[1]
=> [1]
[12] pry(main)> a, = 1, 2; p [a]
[1]
=> [1]
[13] pry(main)> a, = 1, 2, 3; p [a]
[1]
=> [1]
[14] pry(main)> a,b = 1; p [a,b]
[1, nil]
=> [1, nil]
[15] pry(main)> a,b = 1, 2; p [a,b]
[1, 2]
=> [1, 2]
[16] pry(main)> a,b = 1, 2, 3; p [a,b]
[1, 2]
=> [1, 2]
[17] pry(main)> a,b, = 1; p [a,b]
[1, nil]
=> [1, nil]
[18] pry(main)> a,b, = 1, 2; p [a,b]
[1, 2]
=> [1, 2]
[19] pry(main)> a,b, = 1, 2, 3; p [a,b]
[1, 2]
=> [1, 2]
[20] pry(main)> 
[21] pry(main)> proc{|a,| [a]}.call([1])
=> [1]
[22] pry(main)> proc{|a,| [a]}.call([1, 2])
=> [1]
[23] pry(main)> proc{|a,| [a]}.call([1, 2, 3])
=> [1]
[24] pry(main)> proc{|a,b| [a,b]}.call([1])
=> [1, nil]
[25] pry(main)> proc{|a,b| [a,b]}.call([1, 2])
=> [1, 2]
[26] pry(main)> proc{|a,b| [a,b]}.call([1, 2, 3])
=> [1, 2]
[27] pry(main)> proc{|a,b,| [a,b]}.call([1])
=> [1, nil]
[28] pry(main)> proc{|a,b,| [a,b]}.call([1, 2])
=> [1, 2]
[29] pry(main)> proc{|a,b,| [a,b]}.call([1, 2, 3])
=> [1, 2]
[30] pry(main)> 
[31] pry(main)> a, = [1]; p [a]
[1]
=> [1]
[32] pry(main)> a, = [1, 2]; p [a]
[1]
=> [1]
[33] pry(main)> a, = [1, 2, 3]; p [a]
[1]
=> [1]
[34] pry(main)> a,b = [1]; p [a,b]
[1, nil]
=> [1, nil]
[35] pry(main)> a,b = [1, 2]; p [a,b]
[1, 2]
=> [1, 2]
[36] pry(main)> a,b = [1, 2, 3]; p [a,b]
[1, 2]
=> [1, 2]
[37] pry(main)> a,b, = [1]; p [a,b]
[1, nil]
=> [1, nil]
[38] pry(main)> a,b, = [1, 2]; p [a,b]
[1, 2]
=> [1, 2]
[39] pry(main)> a,b, = [1, 2, 3]; p [a,b]
[1, 2]
=> [1, 2]
```
Do you see it? Madness! (If we omit the case of `|x|`) the arguments of `proc` are *exactly* equivalent to `mlhs`. I.e. `proc{|X|}.call(Y)` is exactly the same as `X = Y` for all X and Y. Note that this is not at all true about `lambda`, which works like a real function:

```
[41] pry(main)> lambda{|a,| [a]}.call(1)
=> [1]
[42] pry(main)> lambda{|a,| [a]}.call(1, 2)
ArgumentError: wrong number of arguments (2 for 1)
from (pry):38:in `block in __pry__'
[43] pry(main)> lambda{|a,| [a]}.call(1, 2, 3)
ArgumentError: wrong number of arguments (3 for 1)
from (pry):39:in `block in __pry__'
[44] pry(main)> lambda{|a,b| [a,b]}.call(1)
ArgumentError: wrong number of arguments (1 for 2)
from (pry):40:in `block in __pry__'
[45] pry(main)> lambda{|a,b| [a,b]}.call(1, 2)
=> [1, 2]
[46] pry(main)> lambda{|a,b| [a,b]}.call(1, 2, 3)
ArgumentError: wrong number of arguments (3 for 2)
from (pry):42:in `block in __pry__'
[47] pry(main)> lambda{|a,b,| [a,b]}.call(1)
ArgumentError: wrong number of arguments (1 for 2)
from (pry):43:in `block in __pry__'
[48] pry(main)> lambda{|a,b,| [a,b]}.call(1, 2)
=> [1, 2]
[49] pry(main)> lambda{|a,b,| [a,b]}.call(1, 2, 3)
ArgumentError: wrong number of arguments (3 for 2)
from (pry):45:in `block in __pry__'
[50] pry(main)> 
[51] pry(main)> lambda{|a,| [a]}.call([1])
=> [[1]]
[52] pry(main)> lambda{|a,| [a]}.call([1, 2])
=> [[1, 2]]
[53] pry(main)> lambda{|a,| [a]}.call([1, 2, 3])
=> [[1, 2, 3]]
[54] pry(main)> lambda{|a,b| [a,b]}.call([1])
ArgumentError: wrong number of arguments (1 for 2)
from (pry):49:in `block in __pry__'
[55] pry(main)> lambda{|a,b| [a,b]}.call([1, 2])
ArgumentError: wrong number of arguments (1 for 2)
from (pry):50:in `block in __pry__'
[56] pry(main)> lambda{|a,b| [a,b]}.call([1, 2, 3])
ArgumentError: wrong number of arguments (1 for 2)
from (pry):51:in `block in __pry__'
[57] pry(main)> lambda{|a,b,| [a,b]}.call([1])
ArgumentError: wrong number of arguments (1 for 2)
from (pry):52:in `block in __pry__'
[58] pry(main)> lambda{|a,b,| [a,b]}.call([1, 2])
ArgumentError: wrong number of arguments (1 for 2)
from (pry):53:in `block in __pry__'
[59] pry(main)> lambda{|a,b,| [a,b]}.call([1, 2, 3])
ArgumentError: wrong number of arguments (1 for 2)
from (pry):54:in `block in __pry__'
[60] pry(main)> 
```

Now... what about the "excessed comma" (I think they meant "excess comma") special case? Turns out it's not a special case at all, in fact it is the *normal* case. The special case is having just a single argument, in which case (and this case alone) `proc` works like `lambda`:

```
[61] pry(main)> proc{|a| [a]}.call(1)
=> [1]
[62] pry(main)> proc{|a| [a]}.call(1, 2)
=> [1]
[63] pry(main)> proc{|a| [a]}.call(1, 2, 3)
=> [1]
[64] pry(main)> 
[65] pry(main)> proc{|a| [a]}.call([1])
=> [[1]]
[66] pry(main)> proc{|a| [a]}.call([1, 2])
=> [[1, 2]]
[67] pry(main)> proc{|a| [a]}.call([1, 2, 3])
=> [[1, 2, 3]]
```

The rationale, of course, is that otherwise `[[1],[2]].each do |x|` would do a rather counterintuitive thing, but it's still horrible.

Here's the [file](http://hastebin.com/ucewudixew.rb) I've used for tests, pipe it into irb.

**Comment by whitequark** [comment]
This is actually exactly the semantics [described in compile.c](https://github.com/ruby/ruby/blob/1b94420174641d78b99c45074e8770a91efa2d2b/compile.c#L1466-L1469), by the way.

**Comment by catmando** [comment]
related to #550 I wonder?
